### PR TITLE
Url handling issue 3382

### PR DESCRIPTION
**Pull request by grgrzybek**
This is a tough one, but it's an implementation of my checklist from https://github.com/hawtio/hawtio/issues/3382#issuecomment-2061262976

### CODE REVIEW

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        // 3) not authenticated access to /login - forward to /index.html to get the login page
+        if (loginPage) {
+            redirector.doForward(httpRequest, httpResponse, "/index.html");
+            return;
+        }
+
+        // try pre-emptive authentication, so when user sees index.html page (Hawtio client), jolokia requests
+        // will already be authenticate.
+        // TOCHECK: be carefull with Jolokia/Proxy requests
```
Comment (tadayosi) [inline hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:176]: typos:
- `authenticate` -> `authenticated`
- `carefull` -> `careful`

**Review by tadayosi — APPROVED** [review]
Overall, great fix!  I really like the way you refactor the code as well.

Just a few comments mainly asking for clarification and a few changes.

File: hawtio-system/src/main/java/io/hawt/web/auth/AuthenticationConfiguration.java [hunk 2]
```diff
@@ -25,12 +24,57 @@ public class AuthenticationConfiguration {
...
+    //  - for prefix mapping, like "/jolokia/*", for request like "/jolokia/read/xxx", servlet path is "/jolokia"
+    //  - for extension mapping, like "*.info", for request like "/x/y/z.info", servlet path is ... "/x/y/z.info"
+    //  - for default mapping, "/", for request like "/css/defaults.css", servlet path is "/css/defaults.css"
+
+    // static resources paths, which should always be reachable
+    private static final String[] UNSECURED_RESOURCE_PATHS = {
+            "/index.html", "/favicon.ico", "/hawtconfig.json",
+            "/robots.txt", "/json.worker.js", "/editor.worker.js",
+            "/css", "/fonts", "/img", "/js", "/static"
+    };
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/AuthenticationConfiguration.java:43]: no idea - these are simply there in `hawtio-war/target/hawtio` after you build Hawtio ;)

[the remark this answers is not in the document]
Comment (tadayosi) [reply hawtio-system/src/main/java/io/hawt/web/auth/AuthenticationConfiguration.java:43]: ok, let's leave them until we really know they aren't needed ;-)

File: hawtio-system/src/main/java/io/hawt/web/auth/AuthenticationConfiguration.java [hunk 3]
```diff
@@ -205,7 +249,7 @@ private static List<AuthenticationContainerDiscovery> getDiscoveries(String auth
                 AuthenticationContainerDiscovery discovery = clazz.getDeclaredConstructor().newInstance();
                 discoveries.add(discovery);
             } catch (Exception e) {
-                LOG.warn("Couldn't instantiate discovery " + discoveryClass, e);
+                LOG.warn("Couldn't instantiate discovery {}", discoveryClass, e);
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/AuthenticationConfiguration.java:252]: It may look like this, but implementations do some clever rearrangement to detect throwable among params:
* https://github.com/qos-ch/logback/blob/v_1.5.6/logback-classic/src/main/java/ch/qos/logback/classic/spi/LoggingEvent.java#L141
* https://github.com/apache/logging-log4j2/blob/rel/2.23.1/log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java#L158-L164

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+public class ClientRouteRedirectFilter implements Filter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClientRouteRedirectFilter.class);
+
+    public static final String ATTRIBUTE_UNSECURED_PATHS = "unsecuredPaths";
+
+    private int timeout;
+    private AuthenticationConfiguration authConfiguration;
+
+    // paths which are either unsecured or are secured in different way (like AuthenticationFilter)
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:53]: I usually don't add javadoc to private fields, but sure!

File: hawtio-war/src/main/webapp/WEB-INF/web.xml [hunk 5]
```diff
@@ -245,11 +245,6 @@
     <listener-class>io.hawt.HawtioContextListener</listener-class>
   </listener>
 
-  <error-page>
-    <error-code>404</error-code>
-    <location>/index.html</location>
-  </error-page>
-
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-war/src/main/webapp/WEB-INF/web.xml:252]: This was added by Jimmi in 2014 in 1a5c3e48600abaa87f2bd3a41ac1e490d4890168. Initially I thought it's important, but then I found we're getting proper `index.html` page but with 404 error, which is simply confusing and _may_ affect browser in some future version (like not rendering).

And I really spent last ~2 weeks refactoring this and thinking about it at different angles.
So the reason of this commit was: if user hits CTRL-R, F5 or any kind of browser refresh, current URL (which is ALWAYS a client-side router URL or (somehow) index.html) - and this is what I've carefully handled in LoginRedirectServlet renamed to ClientRouteRedirectServlet.

So you can CTRL-R anything and:
* if it's static resource (image/js/css/index.html/favicon.ico/hawtconfig.json/...), it's served
* it can't be `/jolokia` because it's handled by POST, but if it is (like http://localhost:8080/hawtio/jolokia/version), you'll get JSON
* if it's `/login` (a _special_ client-side route URL), it's handled specially depending on whether authentication is enabled or not and whether you're authenticated or not.

[the remark this answers is not in the document]
Comment (tadayosi) [reply hawtio-war/src/main/webapp/WEB-INF/web.xml:252]: OK, then let's remove this item now.

Note that now we then need to provide our own 404.html because the default 404 page from a web container may expose some sensitive information and the security best practice always advises us to provide our own.

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+        // 1) no authentication or external authentication
+        if (!authConfiguration.isEnabled() || authConfiguration.isExternalAuthenticationEnabled()) {
+            if (loginPage) {
+                // /login should be redirected to "/", so user doesn't see login page blinking
+                redirector.doRedirect(httpRequest, httpResponse, "/");
+            } else {
+                // other client-side router URLs should be forwarded to /index.html
+                // and it should be done here, not with:
+                // 1) <error-code>404</error-code> + <location>/index.html</location> hack for WAR
```
[the remark this answers is not in the document]
Comment (tadayosi) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:148]: le'ts mark it as resolved as enough discussion on this is done.

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+    public static final String ATTRIBUTE_UNSECURED_PATHS = "unsecuredPaths";
+
+    private int timeout;
+    private AuthenticationConfiguration authConfiguration;
+
+    // paths which are either unsecured or are secured in different way (like AuthenticationFilter)
+    private String[] unsecuredPaths;
+    // base path for hawtio. Should be "/" for hawtio.war
+    // and e.g., "/actuator/hawtio" for Spring Boot (configurable)
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:56]: This indeed is _very_ confusing...
* for hawtio.war (or console.war or whatever.war), _context path_ is `/hawtio` (or `/console`, ...)
* for Spring Boot, there's (by default) **no _context path_**, unless you:
  * configure [`server.servlet.context-path`](https://docs.spring.io/spring-boot/docs/3.2.4/reference/html/application-properties.html#application-properties.server.server.servlet.context-path) property when `server.port` and `management.server.port` are the same
  * configure [`management.server.base-path`](https://docs.spring.io/spring-boot/docs/3.2.4/reference/html/application-properties.html#application-properties.actuator.management.server.base-path) property when `management.server.port` is different
* for hawtio.war there's then nothing between _context path_ and some prefix - all URLs defined in `web.xml` are based just on _context path_ (e.g., `/jolokia` is `/hawtio/jolokia`).
* however for Spring Boot, _actuator_ is handled by Spring MVC and _actuator endpoints_ are called by DispatcherServlet using HandlerAdapters. This is another place where some _infix_ may be used:
  * when `server.port` and `management.server.port` are the same, there's only one DispatcherServlet (and one WebServer in fact) which may have non-custom mapping (default is `/`): using [`spring.mvc.servlet.path`](https://docs.spring.io/spring-boot/docs/3.2.4/reference/html/application-properties.html#application-properties.web.spring.mvc.servlet.path) property
  * when `management.server.port` is different, there are two DispatcherServlets and two WebServers, and the management (actuator) part uses own DispatcherServlet - however **there's no equivalent** of `spring.mvc.servlet.path` for Management Server and `/` is always used
* and additionally, only when `management.server.port` is different, there's [`management.endpoints.web.base-path`](https://docs.spring.io/spring-boot/docs/3.2.4/reference/html/application-properties.html#application-properties.actuator.management.endpoints.web.base-path) which defaults to `/actuator`, but may be empty or `/my/special/prefix/for/actuator` ;)

This all affects what is returned by `request.getServletPath()`, `request.getPathInfo()`, `request.getServletContext().getContextPath()` and `request.getRequestURI()`.

For WAR deployment everything was easy - it was enough to call `request.getServletPath()`, but for example with custom DispatcherServlet mapping, servlet path was returning the mapping of DispatcherServlet...

I tried to cover all the cases in `io.hawt.springboot.EndpointPathResolverTest`

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        // "/login" is similar to URLs like "/connect/remote" or "/jmx", because it's client-side router URL
+        // but it shold be handled differently
+        // "/" (and even "") are also treated as client-side routes, because there's no path among
+        // io.hawt.web.auth.AuthenticationConfiguration.UNSECURED_PATHS that "/" startsWith()
+
+        // 1) no authentication or external authentication
+        if (!authConfiguration.isEnabled() || authConfiguration.isExternalAuthenticationEnabled()) {
+            if (loginPage) {
+                // /login should be redirected to "/", so user doesn't see login page blinking
+                redirector.doRedirect(httpRequest, httpResponse, "/");
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:144]: Forward to `/index.html` means simply - render html whatever's the incoming request. This is why we handle all client-side route URLs (like `/jmx`) like this - so user can refresh the page - server gets `/jmx`, responds with index.html content and user is happy.

Before this change, the above scenario was handled by Jimmi's error-page + error-code hack - you got index.html page, but with 404 and I didn't like it ;)

Redirect to `/` is kind of special. If you look closely now - when you're logged in (or authentication is disabled), browsing to `/login` gives you index.html, but react-router immediately analyzes window.location.href and sees `/login`, so client-routes to login page for a moment. I tried to avoid this. That's why the redirect.

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:148]: From SPA perspective it's fine. But returning HTML page with HTTP 404 is just ugly ;)
And this `else` branch does exactly the same (always display HTML), but in better (IMO) way :)

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                redirector.doForward(httpRequest, httpResponse, "/index.html");
+            }
+            return;
+        }
+
+        // 2) when already authenticated - the same as #1, but please leave for clarity
+        if (AuthSessionHelpers.isAuthenticated(session)) {
+            if (loginPage) {
+                // /login should be redirected to "/", so authenticated user doesn't see login page blinking
+                redirector.doRedirect(httpRequest, httpResponse, "/");
```
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:160]: Again - it's about not showing react-router the `/login` route.

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tadayosi) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:56]: Oh, thank you very much for the detailed explanation!  The information is too good to be buried in the review comments... What about adding it to the Javadoc comment?

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tadayosi) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:144]: To be honest, I'm not yet fully convinced that we need this duplication of redirection in both hawtio/hawtio and hawtio-react. But let's accept it for now.

**Comment by tadayosi** [comment]
@grgrzybek What about adding a custom 404.html after we removing the `error-page` entry from web.xml?

File: hawtio-system/src/main/java/io/hawt/util/WebHelper.java
Comment (tadayosi) [inline hawtio-system/src/main/java/io/hawt/util/WebHelper.java:1]: Sorry, but `ServletHelpers` is already there:
https://github.com/hawtio/hawtio/blob/4.x/hawtio-system/src/main/java/io/hawt/web/ServletHelpers.java
which is where I meant at the last comment.

**Comment by grgrzybek** [comment]
oops...

File: hawtio-system/src/main/java/io/hawt/util/WebHelper.java
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/util/WebHelper.java:1]: oh - I didn't find it, let me try again ;)

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java]: See `io.hawt.springboot.EndpointPathResolver#resolve()`

File: hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (grgrzybek) [reply hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java:155]: We can treat it as duplication, but IMO we shouldn't rely on one side (the less secure one) to be the only performing proper navigation.
It's still _less_ duplication than before this PR ;) I could remove some Quarkus and SpringBoot specific redirect/forward mechanisms.
